### ISSUE TITLE
TypedDict errors should report per line to avoid shadowing of multiple errors and confusion

### ISSUE THREAD

**Issue by Daraan** [microsoft/pyright#10156]
**Describe the bug**
Currently annotating a variable as a TypedDict and assigning a dict generates different kinds of errors, depending on weather the assigned dict has only one or two+ errors.

For only one error the error is at the location where the value that is assigned.
If there are two or more errors the whole assignment is counted as an error, this has the following two major problems:
1) If I added an ignore comment for a line that has an error, I unexpectedly silenced all other future errors I might make in the future (compare `one_mistake` vs. `two_mistake_ignored`).
2) If I assign a large dict (`locate_errors` example) with 2+ mistakes the error message quickly becomes very difficult to understand and I do not know which assignments are actually problematic.

It would be great if pyright would treat this more similar to [mypy](https://mypy-play.net/?mypy=latest&python=3.12&gist=f39b7a0ecb439d84834cc89a6366fad0) on a per line bases so nobody traps over the first whole assignment ignore hazard.

**Code or Screenshots**
Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgCYAiSAxjALABQDVANgIYDO7h5AFEQmUo0AlAC4GUSVFajMKOvSlQARrPYwQEqVVnBmYVgoZgUxAPoQk61gGtisguSgBeKAG8tkgESsvUWQCMADSSAMRQCHAgSGgAFjCyMSi4xADaIMQ4eACCnEkQxPL8xAC6YVDJUMQg4CBQsdXE0jQArqzMzHCeUF7KfrIATEMhUOEAPAC0U1U1uJhcyfjsSMyFVGTdXlT9UADMI%2BGR0XEJmGjJGemZuDC5y%2BcFRSRloxVgM7X1jc0wbR1d9AAvgwGDAAO5gCxWGC2ezcFzuTa%2Bfw9YBgMBeA5QMHgdCbPoooYDEZKcI4kxoTbbFH7JTlMA2BjAxj0cGQyzWOxmJIpUgOJyuDyKKQ%2BHZeNEYrFHGLxRLnFJXLK3PIPQowYolfE7IkkqTjabVT6LKAZABuxHaG2F3mpsn2TJB9BYHC4ABlWCA0MRHHwSBRqDAxN0ZHJ8AAfKDqTTWlS6fSGKARqOJqAAORMxG6OlDKfTpm6fKgzGhqVQMDKEd%2BCFWqSjFbTGZTZe6cLLKeTEbzmZZ4QAEh7SLB3uwcPhDbguCYixSoKQA1wAGT1AdD2AgVjrD5zAqcVhe8T0fRUQzmccgdiyd2e70CxExrY7YJInZdoKbUhiiWYrWDAB0AAYHRZBggA)

```python
from typing import TypedDict

class TD(TypedDict):
    a: int
    b: str
    c: float

one_mistake: TD = {
    "a" : 1,  # pyright: ignore[reportAssignmentType]  # no error here actually
    "b" : 222,  # <--- error is not silenced
    "c" : 3,  # pyright: ignore[reportAssignmentType]  # no error here actually
}

two_mistake: TD = {
    "a" : "foo",  # wrong
    "b" : 222,      # wrong
    "c" : 3,      # ok
}

two_mistake_ignored: TD = {
    "a" : "foo",  # pyright: ignore[reportAssignmentType]
    "b" : 222,      # <-- error not revealed
    "c" : 3,
}

class LargeTD(TypedDict):
    a: int | str
    b: float | str | None
    c: int | None
    d: list[int] | tuple[str] | None | int
    e: int | str | None

# Hard to spot errors on long dicts & hard to trace error message:
locate_errors: LargeTD = {
    "c" : 1,
    "a" : None,
    "d" : "foo",
    "b": 2.0
}
```

**VS Code extension or command-line**
1.1.397

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

By default, when the RHS cannot be inferred using bidirectional type inference, pyright falls back on using default inference rules. In the case of a dictionary expression that doesn't conform with a typed dict definition, that means the expression is inferred as a `dict[str, T]` where `T` is a union of all encountered value types. The error is then reported when the `dict[str, T]` is assigned back to the typed dict, which is a type violation. The reported error range is the entire RHS expression.

A while back, I added a feature that provides more specific error range targeting when there is a single incorrect key name in the dictionary. This is modeled after a similar feature in the TypeScript language server. The result is a more specific error in this limited case. I'll note that this feature required significant work to add and maintain, and I don't have plans to change it.

In any case, I don't think it makes sense to change such a case into multiple distinct errors. That would change a single type violation error into a bunch of errors. It would also be a breaking change for pyright users.
